### ISSUE TITLE
Simplify logic to render Site Kit tags (Analytics, AdSense, Tag Manager) and add actions for loading them

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#1862]
The logic for Site Kit to render its tags (done by the `Analytics`, `AdSense` and `Tag_Manager` module classes) is currently somewhat hard to follow, and because it renders tags in various formats (e.g. AMP and non-AMP) while still adding all of these hooks independently, there is some duplicate logic, and this has caused errors before where we e.g. changed a condition in one place but not another.

We should consolidate the logic to whether a certain tag needs to be rendered (regardless of AMP or non-AMP) into one place, and then fire an action like `googlesitekit_init_{module}_tag[_amp]`.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Every module that renders tags should have a single action hooked into `template_redirect`, and in that callback it should run the logic to determine whether a tag needs to be rendered, including checking for and determining the correct ID(s) (e.g. property ID / client ID / container ID).
* If the tag will be rendered, the current actions should be hooked in like they are now. However, all duplicate checks which are now handled by the `template_redirect` hook should be removed from the action callbacks. Furthermore, the tag rendering methods should be modified to receive the respective ID as parameter, having them passed from the logic above where they were already determined.
* If the tag will be rendered, an action `googlesitekit_{module}_init_tag[_amp]` should be fired, with the latter `_amp` suffix present or not based on whether this is an AMP request. The relevant ID should be passed to the action.

## Implementation Brief

* Implement approach based on ACs for `Analytics`, `AdSense`, `Tag_Manager` classes:
    * Add action to `template_redirect` in the `register` method.
    * Check all requirements for rendering the tag there.
    * Conditionally add hooks for non-AMP or AMP variant based on `Context::is_amp()`.
        * Update relevant methods used to print tags to have all checks removed, and have them receive the relevant tracking ID in a method parameter; then pass that parameter from the `register` method where it has been detected.
    * Fire new action hooks as defined in the ACs in the respective non-AMP or AMP clause, passing the tracking ID (property ID / client ID / container ID) to it.
* Merge #1863.

## QA Brief

* All Site Kit tags (Analytics, AdSense, Tag Manager) should be tested in detail to ensure that they are still placed in the frontend like before. For each of the three modules:
    * Regular site (no AMP).
    * AMP Standard Mode (with AMP plugin).
    * AMP Reader Mode (with AMP plugin), in both the AMP and the non-AMP version.
* It should also be checked that each of the tags is _not_ printed when the setting to not place the snippet has been toggled off.

## Changelog entry

* Add new action hooks `googlesitekit_analytics_init_tag`, `googlesitekit_analytics_init_tag_amp`, `googlesitekit_adsense_init_tag`, `googlesitekit_adsense_init_tag_amp`, `googlesitekit_tagmanager_init_tag`, `googlesitekit_tagmanager_init_tag_amp` which fire when the respective tag will be printed for the current request, but _before_ any HTML output has been generated.

**Comment by aaemnnosttv** [comment]
@felixarntz it seems like this may have some overlap with https://github.com/google/site-kit-wp/issues/475.

I see you already have a PR here, but curious if you feel we could simplify things while ensuring more consistency between modules by extracting classes for this?

**Comment by felixarntz** [comment]
@aaemnnosttv I had forgotten about that issue, but still like that approach. How about we make that a follow-up to this one? I think the enhancements have their value standalone already, I'd see it as a stepping stone because now the logic is more consistent, so migrating towards #475 will be more straightforward.

**Comment by aaemnnosttv** [comment]
@felixarntz this mostly looks good to me, just a few questions/suggestions here.

> Every module that renders tags should have a single action hooked into `wp`, and in that callback it should run the logic to determine whether a tag needs to be rendered

Any reason why we should hook into `wp` specifically? I think `template_redirect` might be a more appropriate hook to use as `wp` runs in a few other places other than a front-end HTML response. Ideally there would be an action right before the `template_include` filter (we could even use that filter without changing anything), but otherwise I think `template_redirect` on a later priority is probably the closest action which is most likely to result in an html response before anything is output.

> `googlesitekit_init_{module}_tag[_amp]`

Perhaps `googlesitekit_{module}_init_tag[_amp]` might be more appropriate for the hook name? I think the action belongs more to the module than the plugin, which is what I think of with `googlesitekit_init_*` since we already have `googlesitekit_init`.

**Comment by felixarntz** [comment]
@aaemnnosttv Both SGTM. I've amended the ACs and IB accordingly.

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by cole10up** [comment]
**Tested**

Installed latest SK release candidate, activated setup properly.

Ran a series of functional tests around analytics, tag manager and adsense.

Confirmed all modules functioned properly:

![image](https://user-images.githubusercontent.com/57115907/90109197-51c7ab80-dd00-11ea-9587-9088d34b85c4.png)


Passed QA ✅
